src/module/Provider.sol

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity ^0.8.24;
+
+import {IProvider} from "src/interface/IProvider.sol";
+import {IERC4626} from "src/Common.sol";
+import {MainnetContracts as MC} from "script/Contracts.sol";
+
+/*
+    The Provider fetches state from other contracts.
+*/
+
+contract Provider is IProvider {
+    error UnsupportedAsset(address asset);
+
+    mapping(address => uint256) public rates; // in WETH
+
+    function getRate(address asset) external view override returns (uint256) {
+        if (asset == MC.WETH) {
+            return 1e18;
+        } else if (asset == MC.STETH) {
+            return 1e18;
+        } else if (asset == MC.YNETH) {
+            return IERC4626(MC.YNETH).previewRedeem(1e18);
+        } else if (asset == MC.YNLSDE) {
+            return IERC4626(MC.YNLSDE).previewRedeem(1e18);
+        } else if (asset == MC.BUFFER) {
+            return IERC4626(MC.BUFFER).previewRedeem(1e18);
+        } else if (asset == MC.WSTETH) {
+            return IStETH(MC.STETH).getPooledEthByShares(1e18);
+        } else if (asset == MC.METH) {
+            return IMETH(MC.METH).ratio();
+        } else if (asset == MC.OETH) {
+            return IOETH(MC.OETH).assetToEth(1e18);
+        } else if (asset == MC.RETH) {
+            return IRETH(MC.RETH).getExchangeRate();
+        }
+
+        revert UnsupportedAsset(asset);
+    }
+
+    function otherAssets(address vault, address strategy) public view returns (uint256 assets) {
+        if (strategy == MC.YNETH) {
+            (uint256[] memory withdrawalIndexes, WithdrawalRequest[] memory requests) =
+                IynETHwm(MC.YNETH_WM).withdrawalRequestsForOwner(vault);
+
+            uint256 length = withdrawalIndexes.length;
+            if (length == 0) return 0;
+
+            uint256 ynethRate = IERC4626(MC.YNETH).previewRedeem(1e18);
```

Comment:
here you should actually use requests[i].redemptionRateAtRequestTime instead of ynETHRate.

this will save all the gas here since you already load the request in memory.

Reply:
Likewise for ynLSDe, use requests[i]. redemptionRateAtRequestTime